substrate/validator-sets/pallet/src/lib.rs

```diff
@@ -715,40 +779,96 @@ pub mod pallet {
+      // save & end the current serai session.
+      PreviousParticipants::<T>::set(NetworkId::Serai, current_validators);
+      Self::retire_set(ValidatorSet {
+        network: NetworkId::Serai,
+        session: Self::session(NetworkId::Serai).unwrap(),
+      });
 
       // make a new session and get the next validator set.
       Self::new_session();
 
       // Update Babe and Grandpa
-      let session = prior_serai_session.0 + 1;
-      let validators = prior_serai_participants;
-      let next_validators =
+      let session = Self::session(NetworkId::Serai).unwrap();
+      let next_next_validators =
         Participants::<T>::get(NetworkId::Serai).expect("no Serai participants after new_session");
       Babe::<T>::enact_epoch_change(
         WeakBoundedVec::force_from(
-          validators.iter().copied().map(|(id, w)| (BabeAuthorityId::from(id), w)).collect(),
+          next_validators.iter().copied().map(|(id, w)| (BabeAuthorityId::from(id), w)).collect(),
           None,
         ),
         WeakBoundedVec::force_from(
-          next_validators.into_iter().map(|(id, w)| (BabeAuthorityId::from(id), w)).collect(),
+          next_next_validators.into_iter().map(|(id, w)| (BabeAuthorityId::from(id), w)).collect(),
           None,
         ),
-        Some(session),
+        Some(session.0),
       );
       Grandpa::<T>::new_session(
         true,
-        session,
-        validators.into_iter().map(|(id, w)| (GrandpaAuthorityId::from(id), w)).collect(),
+        session.0,
+        next_validators.into_iter().map(|(id, w)| (GrandpaAuthorityId::from(id), w)).collect(),
       );
     }
 
+    // Since babe & grandpa authorities are the same each session, we can check whichever.
+    // Babe also provides convenient function to check for authorities so we just go with that.
+    fn can_slash_serai_validator(validator: &Public) -> bool {
+      Babe::<T>::is_member(&BabeAuthorityId::from(*validator)) ||
+        Self::previous_participants(NetworkId::Serai)
+          .into_iter()
+          .any(|(id, _)| id.into_inner() == *validator)
+    }
+
+    fn slash_serai_validator(validator: &Public) {
+      let network = NetworkId::Serai;
+      let account = *validator;
+      let mut allocation = Self::allocation((network, account)).unwrap();
+
+      // reduce the current allocation to 0.
+      Self::set_allocation(network, account, Amount(0));
+
+      // prevent any future deallocation.
+      let mut to_unlock_on = Self::session(network).unwrap();
+      // TODO: we don't know in which block the it might be deallocating, so we go through all
+      // but this is ugly.
+      for i in 0 .. 4 {
+        to_unlock_on.0 += i;
+        let pending = Self::take_deallocatable_amount(network, to_unlock_on, account);
+        if pending.is_some() {
+          allocation = allocation.add(pending.unwrap());
+          PendingDeallocations::<T>::set((network, to_unlock_on, account), None);
+        }
+      }
```

Comment:
Because you want to track the removed values, I believe there's a PrefixIterator which takes as it iterates. I'd guess it'd be available via `drain_prefix`, yet I haven't checked.